### DISCUSSION TITLE
[Suggestion] Object extension/instance functions

### DISCUSSION THREAD

**Discussion by emceelovin** [microsoft/CsWin32#694] [Ideas]
For Win32 API calls that all operate on a specific object, could this package generate code that offers extension methods, or even just add them into the class that wraps an object - take an HWND for example - so that there could be a more OO approach? I understand that this would be somewhat of a task, so it was just a suggestion. I absolutely love this generator, so anytime I can think of something to make it better, I figured I'd pitch. 

```cs 
// CreateWindowEx(...)/MoveWindow(...)/SetWindowLongPtr(...)

HWND hwnd = new(); // new(style, class, x, y, width, height, ...) ?

hwnd.CreateWindowEx(...);
hwnd.MoveWindow(x, y, width, height, repaint);
hwnd.SetWindowLongPtr(GWL_EX_STYLE, style);
```

**Comment by AArnott** [comment]
It's a neat idea, but it crosses the line into an ever growing world of how we might automate the generation of a new higher level API over win32, which simply cannot be a goal. There is lots of documentation or examples out there for calling win32 apis as well, and the further we depart from the normal APIs, the more of a novel experience it will be, which may require custom documentation, samples, etc.
Now, there is a balance between these two extremes, and we're fairly satisfied with it. Your extension method idea is one we haven't considered in particular, so we'll stew on this and look at possibilities. It will probably not happen, but I wouldn't write it off completely at this point.
Please keep the ideas and feedback coming.
